### DISCUSSION TITLE
Can't get entiy name on mocked client

### DISCUSSION THREAD

**Discussion by clluiz** [prisma/prisma#18055] [Q&A] [answered]
I'm following this tutorial: https://www.prisma.io/blog/testing-series-1-8eRB5p0Y8o#why-vitest

I can configure everything correctly but in my test scenarios I can't get an "entity" name on my mocked client:

`
    prisma.scheduling.findMany.mockResolvedValue([{
      id: 1,
     ....
    }])
`

I get this error:

![Screenshot_20230222_181002](https://user-images.githubusercontent.com/1070968/220759798-48936e75-f6de-4204-b0e5-2773a13fcb44.png)

The error say that scheduling is undefined. Do I need some additional configuration step to make this work?

**Comment by nurul3101** [comment]
Hi @clluiz 👋 

Can you share how you are mocking your Prisma Client? Is this happening only for scheduling model?

**Comment by clluiz** [comment]
Hi @nurul3101 ,  I mocked the client like this:

```
import { PrismaClient } from '@prisma/client'
import { beforeEach } from 'vitest'
import { mockDeep, mockReset } from 'vitest-mock-extended'

beforeEach(() => {
  mockReset(prisma)
})

const prisma = mockDeep<PrismaClient>()
export default prisma
```

And this problem is happening for all models.

**Reply by nurul3101** [reply]
Hmm, this is exactly similar to what is described in the blog. Is your repo open source?

**Reply by clluiz** [reply]
No. I haven't put this code in a repository yet

**Reply by sabinadams** [reply]
Have you run `npx prisma generate` yet?

**Reply by clluiz** [reply]
Hi @sabinadams, I ran this command only for my development database. T thought it would be not necessary to run it for a mocking.

**Reply by sabinadams** [reply]
Following the steps in the tutorial your test should look something like this:

```ts
//test/sample.test.ts
import { expect, test, vi } from "vitest"; // 👈🏻 Added the `vi` import
import prisma from "../libs/__mocks__/prisma";

vi.mock("../libs/prisma");

test("createUser should return the generated user", async () => {
  prisma.user.create.mockResolvedValue({ name: "somename", id: 1 });
});
```

The `/libs/__mocks__/prisma.ts` should look like:

```ts
import { PrismaClient } from "@prisma/client";
import { beforeEach } from "vitest";
import { mockDeep, mockReset } from "vitest-mock-extended";

beforeEach(() => {
  mockReset(prisma);
});

const prisma = mockDeep<PrismaClient>();
export default prisma;
```

**Comment by clluiz** [comment]
@sabinadams I ran the command npx prisma generate but the error persists. In the intelisense I can see the tenant entity but the methods like findMany, or create are not available for it.

**Comment by sabinadams** [comment, marked as answer]
> @sabinadams I ran the command npx prisma generate but the error persists. In the intelisense I can see the tenant entity but the methods like findMany, or create are not available for it.

Could you provide a repository where the issue occurs? Sounds like some configuration may be missing somewhere but it's hard to suggest a solution without seeing the code 🤔

**Reply by clluiz** [reply]
Hi @sabinadams here it is: https://github.com/clluiz/kairos.

Once you cloned the project:

* Go to backend2 folder;
* Install dependencies with: npm install
* To run the tests execute npm run test
* The file with problems is scheduling.test.ts
* To reproduce just uncomment lines 9 and 10, and 30-38, and comment  from 40-206

**Reply by sabinadams** [reply]
So I noticed a few issues (also couldn't run the project following your instructions initially as there were problems with the app related to the `plugins` folder and the `clearDatabase` function).

In your code you have:

```ts
import prismaMock from '../../prisma/__mocks__/client'
vi.mock('../../prisma/__mocks__/client')
```

The `vi.mock` should be given the filepath of the module you are actually mocking, not the mocked module. So it should instead be:

```diff
import prismaMock from '../../prisma/__mocks__/client'
- vi.mock('../../prisma/__mocks__/client')
+ vi.mock('../../prisma/client')
```

Next, in the actual test you have the following:

```ts
const tenants = prismaMock.tenant.findFirst.mockResolvedValue(
  {
    id: 1,
    name: 'Test',
    logo: 'asda'
  }
)
expect(tenants[0].name).toBe('Test')
```

The `mockResolvedValue` simply lets the mock know what to return next time the function is invoked. It doesn't actually run the function. So rather you should write:

```ts
prismaMock.tenant.findFirst.mockResolvedValue(
  {
    id: 1,
    name: 'Test',
    logo: 'asda'
  }
)

await prisma.tenant.findFirst()
// ...
```

Also, `findFirst` returns a single element (or `NULL`), so the complete test should look like the following:

```diff
prismaMock.tenant.findFirst.mockResolvedValue(
  {
    id: 1,
    name: 'Test',
    logo: 'asda'
  }
)

await prisma.tenant.findFirst()
- expect(tenants[0].name).toBe('Test')
+ expect(tenants?.name).toBe('Test')
```

Hope this helps.

**Comment by clluiz** [comment]
@sabinadams thanks for your help. I made the changes and is working now! I Forgot to include a step to run docker before running the tests.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
